engine/cdn/cdn_item.go

```diff
@@ -34,45 +30,52 @@ func (s *Service) completeItem(ctx context.Context, itemUnit storage.ItemUnit) e
 	item.Status = index.StatusItemCompleted
 
 	var reader io.ReadCloser
-	switch item.Type {
-	case index.TypeItemStepLog, index.TypeItemServiceLog:
+	switch itemUnit.UnitID {
+	case s.Units.Buffer.ID():
 		// Get all data from buffer and add manually last line
 		reader, err = s.Units.Buffer.NewReader(itemUnit)
 		if err != nil {
 			return err
 		}
 	default:
-		return sdk.NewErrorFrom(sdk.ErrInvalidData, "unable to complete item: unknown item type %s", item.Type)
-
+		for _, unit := range s.Units.Storages {
+			if unit.ID() == itemUnit.UnitID {
+				reader, err = unit.NewReader(itemUnit)
+				if err != nil {
+					return err
+				}
+				break
+			}
+		}
+		if reader == nil {
+			return sdk.WithStack(fmt.Errorf("unable to find unit storage %s", itemUnit.ID))
+		}
 	}
 
 	// Compte md5 and sha512
-	md5 := md5.New()
-	sha512 := sha512.New()
+	md5Hash := md5.New()
+	sha512Hash := sha512.New()
 	// For optimum speed, Getpagesize returns the underlying system's memory page size.
 	pagesize := os.Getpagesize()
 	// wraps the Reader object into a new buffered reader to read the files in chunks
 	// and buffering them for performance.
 	mreader := bufio.NewReaderSize(reader, pagesize)
-	multiWriter := io.MultiWriter(md5, sha512)
+	multiWriter := io.MultiWriter(md5Hash, sha512Hash)
 	size, err := io.Copy(multiWriter, mreader)
 	if err != nil {
 		return sdk.WithStack(err)
 	}
 
-	sha512Hash := hex.EncodeToString(sha512.Sum(nil))
-	md5Hash := hex.EncodeToString(md5.Sum(nil))
+	sha512S := hex.EncodeToString(sha512Hash.Sum(nil))
+	md5S := hex.EncodeToString(md5Hash.Sum(nil))
 
-	item.Hash = sha512Hash
-	item.MD5 = md5Hash
+	item.Hash = sha512S
+	item.MD5 = md5S
 	item.Size = size
 
 	if err := index.UpdateItem(ctx, s.Mapper, tx, item); err != nil {
 		return err
 	}
 
-	if err := tx.Commit(); err != nil {
```

Comment:
why removing ?

Reply:
transaction is no more open by this method, so this method does not need commit anymore